src/Tablebot/Plugins/Roll/Dice/DiceStats.hs

```diff
@@ -75,18 +76,63 @@ rangeListValues lv = do
 -- has a variety of  functions that operate on them.
 --
 -- An `Data.Distribution.Experiment` is a monadic form of this.
-class Range a where
+class PrettyShow a => Range a where
   -- | Try and get the `Experiment` of the given value, throwing a
   -- `MonadException` on failure.
   range :: (MonadException m, PrettyShow a) => a -> m Experiment
   range a = propagateException (prettyShow a) (range' a)
 
   range' :: (MonadException m, PrettyShow a) => a -> m Experiment
 
+instance (Range a) => Range (MiscData a) where
+  range' (MiscVar l) = range l
+  range' (MiscIf i) = rangeIfExpr range i
+
+instance (RangeList a) => RangeList (MiscData a) where
+  rangeList' (MiscVar l) = rangeList l
+  rangeList' (MiscIf i) = rangeIfExpr rangeList i
+
+rangeIfExpr :: (MonadException m, Ord b) => (a -> m (D.Experiment b)) -> If a -> m (D.Experiment b)
+rangeIfExpr func (If b t f) = do
+  b' <- range b
+  let mp = toMap $ run b'
+      canBeFalse = M.member 0 mp
+      canBeTrue = M.null $ M.filterWithKey (\k _ -> k /= 0) mp
+      emptyExp = from $ D.fromList @_ @Integer []
+  t' <- if canBeTrue then func t else return emptyExp
+  f' <- if canBeFalse then func f else return emptyExp
+  return $
+    do
+      b'' <- b'
+      if b'' /= 0 then t' else f'
```

Comment:
This _seems_ to weight the false and true cases evenly even if they would not be even - e.g. `if d4-1 then 1 else 0` would (by my theory) give 50% to 1 and 50% to 0. Interestingly however, the actual bot behaves very weirdly - it gave 1 with a 25% chance and then doesn't list 0 as a possible result at all. (So the bars only add up to 25%.)

I would like a little documentation making clear why this _does_ manage to weight things correctly, along with a fix for the result still being wrong.

Screenshot of it at work:
![image](https://user-images.githubusercontent.com/4217801/166311388-5e975754-5b55-4919-9786-420fa54e8cb0.png)

Reply:
So that's 0 with a probability of 25%, which is what would be expected, but the 1 line isn't showing. Just reading this code I can tell what the issue is; the `ifCanBeTrue` is slightly wrong. It filters the distribution of the Boolean values to non zero, and then checks if it's null. If this was correct, it would use not to get the opposite, but alas I wrote this code in a fever dream. 

Easy fix when I get around to it.

Reply:
Why do you think it should be a 50/50 split?